File: algorithms/dfs.java
// Implementation of depth-first-search.
class dfs {
    public static void dfsTraversal(MyNode current){
        if (current == null){
            return;
        }
        System.out.print(current.data + " ");
        dfsTraversal(current.left);
        dfsTraversal(current.right);
    }

    public static MyNode createTree(){
        MyNode root = new MyNode(10, new MyNode(6, 
                                 new MyNode(3), new MyNode(8)),
                                 new MyNode(13, 
                                 new MyNode(11), null));

        return root;
    }

    public static void main(String[] args){
        MyNode root = createTree();
        System.out.print("DFS: ");
        dfsTraversal(root); // DFS: 10 6 3 8 13 11 
    }
}

File: algorithms/dfs.py
def dfs_recursive(graph, vertex, path = []):
    """Perform depth-first-search recursively
    
    Arguments:
        graph {dict} -- Graph data structure implemented as a dict of sets.
    """
    path += [vertex]
    for node in graph[vertex]:
        if node not in path:
            path = dfs_recursive(graph, node, path)
    return path

def dfs_iterative(graph, start):
    """Perform depth-first-search iteratively
    
    Arguments:
        graph {dict} -- Graph data structure implemented as a dict of sets.
    """
    stack, path = [start], []

    while stack:
        vertex = stack.pop()
        if vertex in path:
            continue # jump to next for loop
        path.append(vertex)
        for node in graph[vertex]:
            stack.append(node)
    
    return path


if __name__ == '__main__':
    graph = {'A': set(['B', 'C']),
         'B': set(['A', 'D', 'E']),
         'C': set(['A', 'F']),
         'D': set(['B']),
         'E': set(['B', 'F']),
         'F': set(['C', 'E'])}
    print(dfs_recursive(graph, 'A'))
    print(dfs_iterative(graph, 'A'))